doc/fsarchiver.8
.TH FSARCHIVER 8 "30 December 2009"

.SH NAME
fsarchiver \- filesystem archiver

.SH DESCRIPTION
fsarchiver is a system tool that allows you to save the contents of a 
filesystem to a compressed archive file. The file-system can be restored 
on a partition which has a different size and it can be restored on a 
different file-system. Unlike tar/dar, FSArchiver also creates the 
filesystem when it extracts the data to partitions. Everything is 
checksummed in the archive in order to protect the data. If the archive 
is corrupt, you just lose the current file, not the whole archive. 

.SH LINKS
.SS Official project homepage:
http://www.fsarchiver.org
.SS Quick Start Guide:
http://www.fsarchiver.org/QuickStart
.SS Forums where to ask questions:
http://www.fsarchiver.org/forums/
.SS Report a bug:
http://www.fsarchiver.org/forums/viewforum.php?f=17

.SH SYNOPSIS
.B fsarchiver [
.I options
.B ] savefs
.I archive
.I filesystem
.B ...
.PP
.B fsarchiver [
.I options
.B ] restfs
.I archive
.BI id= n ,dest= filesystem [,mkfs= fstype, mkfsopt= options ]
.B ...
.PP
.B fsarchiver [
.I options
.B ] savedir
.I archive
.I directory
.B ...
.PP
.B fsarchiver [
.I options
.B ] restdir
.I archive destination
.PP
.B fsarchiver [
.I options
.B ] archinfo
.I archive
.PP
.B fsarchiver [
.I options
.B ] probe [detailed]

.SH COMMANDS
.TP
.B savefs
Save
.I filesystems
to
.IR archive .
.TP
.B restfs
Restore filesystems from
.IR archive .
This overwrites the existing data on
.IR filesystems .
Zero-based index
.I n
indicates the part of the archive to restore.
Optionally, a filesystem may be converted to
.IR fstype .
.TP
.B savedir
Save
.I directories
to
.I archive
(similar to a compressed tarball).
.TP
.B restdir
Restore data from 
.I archive
which is not based on a filesystem to
.IR destination .
.TP
.B archinfo
Show information about an existing
.I archive
file and its contents.
.TP
.B probe
Show list of filesystems detected on the disks.



.SH "OPTIONS"
.PP
.IP "\fB\-h, \-\-help\fP"
Show help and information about how to use fsarchiver with examples.
.IP "\fB\-V, \-\-version\fP"
Show program version and exit.
.IP "\fB\-v, \-\-verbose\fP"
Verbose mode (can be used several times to increase the level of details).
The details will be printed to the console.
.IP "\fB\-o, \-\-overwrite\fP"
Overwrite the archive if it already exists instead of failing.
.IP "\fB\-d, \-\-debug\fP"
Debug mode (can be used several times to increase the level of details).
The details will be written in /var/log/fsarchiver.log.
.IP "\fB\-A, \-\-allow-rw-mounted\fP"
Allow to save a filesystem which is mounted in read-write (live backup).
By default fsarchiver fails with an error if the partition if mounted 
in read-write mode which allows modifications to be done on the filesystem
during the backup. Modifications can drive to inconsistencies in the backup.
Using lvm snapshots is the recommended way to make backups since it will
provide consistency, but it is only available for filesystems which are on
LVM logical-volumes.
.IP "\fB\-a, \-\-allow-no-acl-xattr\fP"
Allow to run savefs when partition is mounted without the acl/xattr options.
By default fsarchiver fails with an error if the partition is mounted
in such a way that the ACL and Extended-Attributes are not readable. These
attributes would not be saved and then such attributes could be lost. If you
know what you don't need ACL and Extended-Attributes to be preserved then
it's safe to run fsarchiver with that option.
.IP "\fB\-e pattern, \-\-exclude=pattern\fP"
Exclude files and directories that match that pattern. The pattern can 
contains shell asterisks such as * and ?, and the pattern may be either
a simple file/dir name or an absolute file/dir path. You must use quotes
around the pattern each time you use wildcards, else it would be interpreted
by the shell. The wildcards must be interpreted by fsarchiver. See examples
below for more details about this option.
.IP "\fB\-L label, \-\-label=label\fP"
Set the label of the archive: it's just a comment about the contents. 
It can be used to remember a particular thing about the archive or the
state of the filesystem for instance.
.IP "\fB\-y level, \-\-ecclevel=level\fP"
Valid error correction levels are between 0 (no error correction)
and 16 (very good error correction). Using Error correction means it
will write extra packets in the archive file to allow fsarchiver to fix
the archive automatically if a corruption is found when you attempt to
restore it. With no error correction, fsarchiver won't be able to fix it
and you will loose the current block, so the current file will be lost.
If you use a low level of error correction (eg: 1, 2, ...)  then the 
archive will be a little bigger and fsarchiver will be able to automatically
fix corruptions in the archive if there are not too many corrupt packets
in the archive. The higher the error correction level is the bigger the 
archive will be and the better is the ability to fix important corruptions
in the archive. Fixing a corruption means being actually able to restore
100% of your original data (files and directories) without any error in it.
.IP "\fB\-z level, \-\-compress=level\fP"
Valid compression levels are between 1 (very fast) and 9 (very good).
The memory requirement increases a lot with the best compression levels,
and it's multiplied by the number of compression threads (option -j).
Level 9 is considered as an extreme compression level and requires an
huge amount of memory to run.
For more details please read this page: http://www.fsarchiver.org/Compression
.IP "\fB\-s mbsize, \-\-split=mbsize\fP"
Split the archive into several files of mbsize megabytes each.
.IP "\fB\-j count, \-\-jobs=count\fP"
Create more than one compression thread. Useful on multi-core CPUs. 
By default fsarchiver will only use one compression thread (-j 1) and
then only one logical processor will be used for compression.
You should use that option if you have a multi-core CPU or more than one
physical CPU on your computer. The typical way to use this option is to
specify the number of logical processors available so that all the
processing power is used to compress the archive very quickly. You may 
also want to use all the logical processors but one for that task so that
the system stays responsive for other applications.
.IP "\fB\-c password, \-\-cryptpass=password\fP"
Encrypt/decrypt data in archive. Password length: 6 to 64 chars.
You can either provide a real password or a dash ("-c -") with this option
if you do not want to provide the password in the command line and you

ChangeLog
fsarchiver: Filesystem Archiver for Linux [http://www.fsarchiver.org]
=====================================================================
* 0.7.0:
  - Major feature: implemented savept, restpt, showpt (save and restore partition tables)
  - Added configuration file: /etc/fsarchiver.conf (it currently only supports "exclude_restpt")
  - Implemented error correction based on Forward Error Correction (FEC) (cf: option -y)
  - File format more robust when corruptions / insertions / deletions happen in the archive
* 0.6.17:
  - Implemented "mkfsopt" restfs option to pass extra options to mkfs (Michael Moninski)
  - Fixed parsing of "/proc/self/mountinfo" (mount options were not parsed on new systems)
* 0.6.16 (2013-02-07):
  - Fixed mke2fs requirement for the "ext_attr" feature (affects RHLE5/CentOS5)
  - Fixed parsing of "/proc/self/mountinfo" for systemd based systems
* 0.6.15 (2012-06-02):
  - Added support for recent ext4 features (up to linux-3.4 / e2fsprogs-1.42)
* 0.6.14 (2012-06-02):
  - Added support for recent btrfs features (up to linux-3.4)
  - Fixed compilation warnings (variable ‘res’ set but not used)
* 0.6.13 (2012-03-04):
  - Fixed detection of the root filesystem using "/proc/self/mountinfo" instead of "/proc/mounts"
* 0.6.12 (2010-12-25):
  - Fix: get correct mount info for root device when not listed in /proc/mounts (eg: missing "/dev/root")
* 0.6.11 (2010-12-01):
  - Updated supported btrfs compat flags to make it work with btrfs-2.6.35
  - The -c/--cryptpass option now supports interactive passwords: use "-c -"
* 0.6.10 (2010-05-09):
  - Fixed support of symbolic links on ntfs filesystems with ntfs3g >= 2010.3.6
* 0.6.9 (2010-05-04):
  - Fix in probe: show devmapper/lvm volumes even when /dev/dm-xx does not exist
  - Fixed restoration of very small archives (archive < 4K)
  - Fix error handling in restoration: consider ENOSPC as a fatal error and other fixes
  - Fix: remove all volumes of the archive instead of just the first one if save fails
* 0.6.8 (2010-02-20):
  - Fixed compilation error on systems with recent kernel headers by including <sys/stat.h>
  - Fixed critial bug: there was a risk of corruption when the archive was written on a smbfs/cifs filesystem
* 0.6.7 (2010-01-31):
  - Added support for sparse files (sparse file on the original disk will be recreated as sparse file)
  - Added per-archive minimum fsarchiver version requirement (MAINHEADKEY_MINFSAVERSION)
  - Added dirsinfo in archives with simple files and directories to store stats required for progression
  - The logfile created when option -d is used now has a specific name so that it's not overwritten
  - The md5 checksums are now calculated using the implementation from libgcrypt instead of the internal one
  - The libgcrypt library is now a mandatory dependency and crypto cannot be disabled any more
  - Allow non root users to use "fsarchiver archinfo" as long as they have read permissions on the archive
* 0.6.6 (2010-01-24):
  - Fix: don't remove the archive file when savefs/savedir fails because the archive already exists
  - Partitions already mounted are remounted with MS_BIND to have access to files hidden by mounted filesystems 

src/fsarchiver.c
/*
 * fsarchiver: Filesystem Archiver
 *
 * Copyright (C) 2008-2012 Francois Dupoux.  All rights reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License v2 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * Homepage: http://www.fsarchiver.org
 */

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif

#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <getopt.h>
#include <stdlib.h>
#include <assert.h>

#include "fsarchiver.h"
#include "dico.h"
#include "common.h"
#include "restore.h"
#include "save.h"
#include "probe.h"
#include "archinfo.h"
#include "syncthread.h"
#include "comp_lzo.h"
#include "crypto.h"
#include "options.h"
#include "logfile.h"
#include "error.h"
#include "queue.h"
#include "iobuffer.h"
#include "archio.h"
#include "layout_save.h"
#include "layout_rest.h"

u32 g_valid_header_types[] = {FSA_HEADTYPE_MAIN, FSA_HEADTYPE_FSIN,
    FSA_HEADTYPE_FSYB, FSA_HEADTYPE_DIRS, FSA_HEADTYPE_OBJT,
    FSA_HEADTYPE_BLKH, FSA_HEADTYPE_FILF, FSA_HEADTYPE_DATF, 0};

char g_archive[PATH_MAX];
int g_archver = FSA_FMT_NULL;

void usage(char *progname, bool examples)
{
    int lzo=false, lzma=false;

#ifdef OPTION_LZO_SUPPORT
    lzo=true;
#endif // OPTION_LZO_SUPPORT
#ifdef OPTION_LZMA_SUPPORT
    lzma=true;
#endif // OPTION_LZMA_SUPPORT

    msgprintf(MSG_FORCE, "====> fsarchiver version %s (%s) - http://www.fsarchiver.org <====\n", FSA_VERSION, FSA_RELDATE);
    msgprintf(MSG_FORCE, "Distributed under the GPL v2 license (GNU General Public License v2).\n");
    msgprintf(MSG_FORCE, " * usage: %s [<options>] <command> <archive> [<part1> [<part2> [...]]]\n", progname);
    msgprintf(MSG_FORCE, "<commands>\n");
    msgprintf(MSG_FORCE, " * savefs: save filesystems to an archive file (backup a partition to a file)\n");
    msgprintf(MSG_FORCE, " * restfs: restore filesystems from an archive (overwrites the existing data)\n");
    msgprintf(MSG_FORCE, " * savedir: save directories to the archive (similar to a compressed tarball)\n");
    msgprintf(MSG_FORCE, " * restdir: restore data from an archive which is not based on a filesystem\n");
    msgprintf(MSG_FORCE, " * savept: save partition tables of the local disks to an archive (safe)\n");
    msgprintf(MSG_FORCE, " * restpt: restore partition table of a local disk from an archive (dangerous)\n");
    msgprintf(MSG_FORCE, " * showpt: show partition tables which have been save in an archive\n");
    msgprintf(MSG_FORCE, " * archinfo: show information about an existing archive file and its contents\n");
    msgprintf(MSG_FORCE, " * probe [-v]: show list of local disks and filesystems\n");
    msgprintf(MSG_FORCE, "<options>\n");
    msgprintf(MSG_FORCE, " -o: overwrite the archive if it already exists instead of failing\n");
    msgprintf(MSG_FORCE, " -v: verbose mode (can be used several times to increase the level of details)\n");
    msgprintf(MSG_FORCE, " -d: debug mode (can be used several times to increase the level of details)\n");
    msgprintf(MSG_FORCE, " -A: allow to save a filesystem which is mounted in read-write (live backup)\n");
    msgprintf(MSG_FORCE, " -a: allow running savefs when partition mounted without the acl/xattr options\n");
    msgprintf(MSG_FORCE, " -N: do not save a copy the partition tables of local disks during the savefs\n");
    msgprintf(MSG_FORCE, " -e <pattern>: exclude files and directories that match that pattern\n");
    msgprintf(MSG_FORCE, " -L <label>: set the label of the archive (comment about the contents)\n");
    msgprintf(MSG_FORCE, " -y <level>: error-correction level from %d (no correction) to %d (best) default=%d\n", FSA_MIN_ECCLEVEL, FSA_MAX_ECCLEVEL, FSA_DEF_ECCLEVEL);
    msgprintf(MSG_FORCE, " -z <level>: compression level from 1 (very fast)  to  9 (very good) default=3\n");
    msgprintf(MSG_FORCE, " -s <mbsize>: split the archive into several files of <mbsize> megabytes each\n");
    msgprintf(MSG_FORCE, " -j <count>: create more than one compression thread. useful on multi-core cpu\n");
    msgprintf(MSG_FORCE, " -c <password>: encrypt/decrypt data in archive, \"-c -\" for interactive password\n");
    msgprintf(MSG_FORCE, " -h: show help and information about how to use fsarchiver with examples\n");
    msgprintf(MSG_FORCE, " -V: show program version and exit\n");
    msgprintf(MSG_FORCE, "<information>\n");
    msgprintf(MSG_FORCE, " * Support included for: lzo=%s, lzma=%s\n", (lzo==true)?"yes":"no", (lzma==true)?"yes":"no");
    msgprintf(MSG_FORCE, " * fsarchiver is still in development, don't use it for production.\n");
    
    if (examples==true)
    {
        msgprintf(MSG_FORCE, "<examples>\n");
        msgprintf(MSG_FORCE, " * \e[1msave only one filesystem (/dev/sda1) to an archive:\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver savefs /data/myarchive1.fsa /dev/sda1\n");
        msgprintf(MSG_FORCE, " * \e[1msave two filesystems (/dev/sda1 and /dev/sdb1) to an archive:\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver savefs /data/myarchive2.fsa /dev/sda1 /dev/sdb1\n");
        msgprintf(MSG_FORCE, " * \e[1mrestore the first filesystem from an archive (first = number 0):\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver restfs /data/myarchive2.fsa id=0,dest=/dev/sda1\n");
        msgprintf(MSG_FORCE, " * \e[1mrestore the second filesystem from an archive (second = number 1):\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver restfs /data/myarchive2.fsa id=1,dest=/dev/sdb1\n");
        msgprintf(MSG_FORCE, " * \e[1mrestore two filesystems from an archive (number 0 and 1):\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver restfs /data/arch2.fsa id=0,dest=/dev/sda1 id=1,dest=/dev/sdb1\n");
        msgprintf(MSG_FORCE, " * \e[1mrestore a filesystem from an archive and convert it to reiserfs:\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver restfs /data/myarchive1.fsa id=0,dest=/dev/sda1,mkfs=reiserfs\n");
        msgprintf(MSG_FORCE, " * \e[1msave the contents of /usr/src/linux to an archive (similar to tar):\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver savedir /data/linux-sources.fsa /usr/src/linux\n");
        msgprintf(MSG_FORCE, " * \e[1msave a filesystem (/dev/sda1) to an archive split into volumes of 680MB:\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver savefs -s 680 /data/myarchive1.fsa /dev/sda1\n");
        msgprintf(MSG_FORCE, " * \e[1msave a filesystem and exclude all files/dirs called 'pagefile.*':\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver savefs /data/myarchive.fsa /dev/sda1 --exclude='pagefile.*'\n");
        msgprintf(MSG_FORCE, " * \e[1mgeneric exclude for 'share' such as '/usr/share' and '/usr/local/share':\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver savefs /data/myarchive.fsa --exclude=share\n");
        msgprintf(MSG_FORCE, " * \e[1mabsolute exclude valid for '/usr/share' but not for '/usr/local/share':\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver savefs /data/myarchive.fsa --exclude=/usr/share\n");
        msgprintf(MSG_FORCE, " * \e[1msave a filesystem (/dev/sda1) to an encrypted archive:\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver savefs -c mypassword /data/myarchive1.fsa /dev/sda1\n");
        msgprintf(MSG_FORCE, " * \e[1mSame as before but prompt for password in the terminal:\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver savefs -c - /data/myarchive1.fsa /dev/sda1\n");
        msgprintf(MSG_FORCE, " * \e[1mextract an archive made of simple files to /tmp/extract:\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver restdir /data/linux-sources.fsa /tmp/extract\n");
        msgprintf(MSG_FORCE, " * \e[1mshow information about an archive and its file systems:\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver archinfo /data/myarchive2.fsa\n");
        msgprintf(MSG_FORCE, " * \e[1msave description of the partition tables of local disks:\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver savept /data/myarchive.fsa\n");
        msgprintf(MSG_FORCE, " * \e[1mshow partition tables descriptions saved in an archive file:\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver showpt /data/myarchive.fsa\n");
        msgprintf(MSG_FORCE, " * \e[1mrecreate partition table on /dev/sdb from first description saved:\e[0m\n");
        msgprintf(MSG_FORCE, "   fsarchiver restpt /data/myarchive.fsa id=0,dest=/dev/sdb\n");
    }
}

static struct option const long_options[] =
{
    {"overwrite", no_argument, NULL, 'o'},
    {"allow-no-acl-xattr", no_argument, NULL, 'a'},
    {"allow-rw-mounted", no_argument, NULL, 'A'},
    {"nosavept", no_argument, NULL, 'N'},
    {"verbose", no_argument, NULL, 'v'},
    {"debug", no_argument, NULL, 'd'},
    {"ecclevel", required_argument, NULL, 'y'},
    {"compress", required_argument, NULL, 'z'},
    {"jobs", required_argument, NULL, 'j'},
    {"help", no_argument, NULL, 'h'},
    {"version", no_argument, NULL, 'V'},
    {"split", required_argument, NULL, 's'},
    {"cryptpass", required_argument, NULL, 'c'},
    {"label", required_argument, NULL, 'L'},
    {"exclude", required_argument, NULL, 'e'},
    {NULL, 0, NULL, 0}
};

int process_cmdline(int argc, char **argv)
{
    bool runasroot=true;
    char *probemode;
    sigset_t mask_set;
    bool probedetailed=0;
    char *command=NULL;
    char tempbuf[1024];
    char *progname;
    int argcok;
    int ret=0;
    int cmd;
    int c;
    
    // init
    progname=argv[0];
    
    // set default options
    g_options.overwrite = false;
    g_options.allowsaverw = false;
    g_options.dontcheckmountopts = false;
    g_options.nosavept = false;
    g_options.verboselevel = 0;
    g_options.debuglevel = 0;
    g_options.compressjobs = 1;
    g_options.ecclevel = FSA_DEF_ECCLEVEL;
    g_options.fsacomplevel = 3; // fsa level 3  =  "gzip -6"
    g_options.compressalgo = FSA_DEF_COMPRESS_ALGO;
    g_options.compresslevel = FSA_DEF_COMPRESS_LEVEL; // default level for gzip
    g_options.datablocksize = FSA_DEF_BLKSIZE;
    g_options.encryptalgo = ENCRYPT_NONE;
    snprintf(g_options.archlabel, sizeof(g_options.archlabel), "<none>");
    g_options.encryptpass[0] = 0;
    
    while ((c = getopt_long(argc, argv, "oaAvdy:z:j:hVs:c:L:e:N", long_options, NULL)) != EOF)
    {
        switch (c)
        {
            case 'o': // overwrite existing archives
                g_options.overwrite=true;
                break;
            case 'a': // don't check the mount options for already-mounted filesystems
                g_options.dontcheckmountopts=true;
                break;
            case 'A': // allows to backup read/write mounted partition
                g_options.allowsaverw=true;
                break;
            case 'N': // do not save partition tables
                g_options.nosavept=true;
                break;
            case 'v': // verbose mode
                g_options.verboselevel++;
                break;
            case 'V': // version
                msgprintf(MSG_FORCE, "fsarchiver %s (%s)\n", FSA_VERSION, FSA_RELDATE);
                return 0;
            case 'd': // debug mode
                g_options.debuglevel++;
                break;
            case 'j': // compression jobs
                g_options.compressjobs=atoi(optarg);
                if (g_options.compressjobs<1 || g_options.compressjobs>FSA_MAX_COMPJOBS)
                {
                    errprintf("[%s] is not a valid number of jobs. Must be between 1 and %d\n", optarg, FSA_MAX_COMPJOBS);
                    usage(progname, false);
                    return 1;
                }
                break;
            case 'e': // exclude files/directories
                strlist_add(&g_options.exclude, optarg);
                break;
            case 's': // split archive into several volumes
                g_options.splitsize=((u64)atoll(optarg))*((u64)1024LL*1024LL);
                if (g_options.splitsize==0)
                {
                    errprintf("argument of option -s is invalid (%s). It must be a valid integer\n", optarg);
                    usage(progname, false);
                    return -1;
                }
                else // show the calculated size since it was probably incorrect before due to an integer overflow
                {
                    msgprintf(MSG_FORCE, "Archive will be split into volumes of %lld bytes (%s)\n",
                        (long long)g_options.splitsize, format_size(g_options.splitsize, tempbuf, sizeof(tempbuf), 'h'));
                }
                break;
            case 'y': // fec correction level
                g_options.ecclevel = atoi(optarg);
                if ((g_options.ecclevel < FSA_MIN_ECCLEVEL) || (g_options.ecclevel > FSA_MAX_ECCLEVEL))
                {
                    errprintf("[%s] is not a valid error-correction-code level, it must be an "
                        "integer between %d and %d.\n", optarg, FSA_MIN_ECCLEVEL, FSA_MAX_ECCLEVEL);
                    usage(progname, false);
                    return -1;
                }
                break;
            case 'z': // compression level
                g_options.fsacomplevel = atoi(optarg);
                if ((g_options.fsacomplevel < 1) || (g_options.fsacomplevel > 9))
                {
                    errprintf("[%s] is not a valid compression level, it must be an integer between 1 and 9.\n", optarg);
                    usage(progname, false);
                    return -1;
                }
                if (options_select_compress_level(g_options.fsacomplevel) < 0)
                    return -1;
                if (g_options.fsacomplevel >= 8)
                    msgprintf(MSG_FORCE, "Compression levels >= 8 may require a huge amount of memory\n"
                        "Please read the man page or \"http://www.fsarchiver.org/Compression\" for more details.\n");
                break;
            case 'c': // encryption
                g_options.encryptalgo = ENCRYPT_BLOWFISH;
                if (((strlen(optarg) < FSA_MIN_PASSLEN) || (strlen(optarg) > FSA_MAX_PASSLEN)) && (strcmp(optarg, "-") != 0))
                {
                    errprintf("the password lenght is incorrect, it must between %d and %d chars, or \"-\" for interactive password prompt.\n", FSA_MIN_PASSLEN, FSA_MAX_PASSLEN);
                    usage(progname, false);
                    return -1;
                }
                snprintf((char*)g_options.encryptpass, FSA_MAX_PASSLEN, "%s", optarg);
                break;
            case 'L': // archive label
                snprintf(g_options.archlabel, sizeof(g_options.archlabel), "%s", optarg);
                break;
            case 'h': // help
                usage(progname, true);
                return 0;
            default:
                usage(progname, false);
                return -1;
        }
    }
    
    argc -= optind;
    argv += optind;
    
    // in all cases we need at least 1 parameters
    if (argc < 1)
    {
        fprintf(stderr, "No arguments provided, cannot continue\n");
        usage(progname, false);
        return -1;
    }
    else // mandatory and unique parameters
    {
        command=*argv++, argc--;
    }
    
    // calculate threshold for small files that are compressed together
    g_options.smallfilethresh=min(g_options.datablocksize/4, FSA_MAX_SMALLFILESIZE);
    
    // convert commands to integers
    if (strcmp(command, "savefs")==0)
    {
        cmd=OPER_SAVEFS;
        runasroot=true;
        argcok=(argc>=2);
    }
    else if (strcmp(command, "restfs")==0)
    {
        cmd=OPER_RESTFS;
        runasroot=true;
        argcok=(argc>=2);
    }
    else if (strcmp(command, "savedir")==0)
    {
        cmd=OPER_SAVEDIR;
        runasroot=true;
        argcok=(argc>=2);
    }
    else if (strcmp(command, "restdir")==0)
    {
        cmd=OPER_RESTDIR;
        runasroot=true;
        argcok=(argc==2);
    }
    else if (strcmp(command, "archinfo")==0)
    {   cmd=OPER_ARCHINFO;
        runasroot=false;
        argcok=(argc==1);
    }
    else if (strcmp(command, "probe")==0)
    {
        cmd=OPER_PROBE;
        runasroot=true;
        argcok=(argc<=1);
    }
    else if (strcmp(command, "savept")==0)
    {
        cmd=OPER_SAVEPT;
        runasroot=true;
        argcok=(argc==1);
    }
    else if (strcmp(command, "restpt")==0)
    {
        cmd=OPER_RESTPT;
        runasroot=true;
        argcok=(argc>=2);
    }
    else if (strcmp(command, "showpt")==0)
    {
        cmd=OPER_SHOWPT;
        runasroot=false;
        argcok=(argc==1);
    }
    else // command not found
    {
        errprintf("[%s] is not a valid command.\n", command);
        usage(progname, false);
        return -1;
    }
    
    // check there are enough parameters on the cmd line
    if (argcok!=true)
    {
        errprintf("invalid number of arguments.\n");
        usage(progname, false);
        return -1;
    }
    
    // check if must be run as root
    if (runasroot==true && geteuid()!=0)
    {
        errprintf("\"fsarchiver %s\" must be run as root. cannot continue.\n", command);
        return -1;
    }
    
    // interactive password
    if (strcmp((char*)g_options.encryptpass, "-")==0)
    {
        int passconfirm;
        char *passtmp=NULL;
         
        passconfirm = (cmd==OPER_SAVEFS || cmd==OPER_SAVEDIR);
        if ((passtmp=getpass("Enter password: "))==NULL)
        {
            errprintf("failed to get interactive password from the console\n");
            return -1;
        }
        if (strlen(passtmp)<FSA_MIN_PASSLEN || strlen(passtmp)>FSA_MAX_PASSLEN)
        {
            errprintf("the password lenght is incorrect, it must between %d and %d chars\n", FSA_MIN_PASSLEN, FSA_MAX_PASSLEN);
            return -1;
        }
        snprintf((char*)g_options.encryptpass, FSA_MAX_PASSLEN, "%s", passtmp);
        
        if (passconfirm==true)
        {
            if ((passtmp=getpass("Confirm password: "))==NULL)
            {
                errprintf("failed to get interactive password from the console\n");
                return -1;
            }
            if (strcmp((char*)g_options.encryptpass, passtmp)!=0)
            {
                errprintf("the passwords do not match\n");
                return -1;
            }
        }
    }
    
    // commands that require an archive as the first argument
    switch (cmd)
    {
        case OPER_SAVEFS:
        case OPER_RESTFS:
        case OPER_SAVEDIR:
        case OPER_RESTDIR:
        case OPER_ARCHINFO:
        case OPER_SAVEPT:
        case OPER_RESTPT:
        case OPER_SHOWPT:
            snprintf(g_archive, PATH_MAX, *argv);
            argv++;
            argc--;
            fprintf(stderr, "This is a development version: the file format is unstable and will change in the future.\n");
            break;
        case OPER_PROBE:
            probedetailed=false; // do "simple" by default
            if (g_options.verboselevel>0)
                probedetailed=true;
            if (argc>0)
            {   probemode=*argv++, argc--;
                if (strcmp(probemode, "detailed")==0)
                    probedetailed=true;
            }
            break;
    }
    
    // list of partitions to backup/restore
    /*for (fscount=0; (fscount < argc) && (argv[fscount]); fscount++)
        partition[fscount]=argv[fscount];*/
    
    // install signal handlers
    sigemptyset(&mask_set);
    sigaddset(&mask_set, SIGINT);
    sigaddset(&mask_set, SIGTERM);
    sigprocmask(SIG_SETMASK, &mask_set, NULL);
    
    if (g_options.debuglevel>0)
        logfile_open();
    
    switch (cmd)
    {
        case OPER_SAVEFS:
        case OPER_SAVEPT:
            ret=save(argc, argv, ARCHTYPE_FILESYSTEMS);
            break;
        case OPER_SAVEDIR:
            ret=save(argc, argv, ARCHTYPE_DIRECTORIES);
            break;
        case OPER_RESTFS:
        case OPER_RESTDIR:
        case OPER_ARCHINFO:
        case OPER_RESTPT:
        case OPER_SHOWPT:
            ret=restore(argc, argv, cmd);
            break;
        case OPER_PROBE:
            ret=oper_probe(probedetailed);
            break;
        default:
            errprintf("[%s] is not a valid command.\n", command);
            usage(progname, false);
            ret=-1;
            break;
    };
    
    logfile_close();
    
    return ret;
}

#include "serializer.h"

int main(int argc, char **argv)
{
    int ret;
    
    /*cserializer *serial1;
    cserializer *serial2;
    s64 number1, number2;
    char string1[512], string2[512];
    char bigbuffer1[8192];
    char bigbuffer2[8192];
    memset(bigbuffer1, 0, sizeof(bigbuffer1));
    memset(bigbuffer2, 0, sizeof(bigbuffer2));
    serial1=serializer_alloc();
    serial2=serializer_alloc();
    serializer_setbykeys_format(serial1, "disk", "device", "/dev/sda");
    serializer_setbykeys_format(serial1, "disk", "serial", "XXXXXXXX");
    serializer_setbykeys_format(serial1, "disk", "partcount", "%d", 3);
    serializer_setbykeys_format(serial1, "part001", "name", "boot");
    serializer_setbykeys_format(serial1, "part001", "size", "%d", 150000);
    serializer_setbykeys_format(serial1, "part002", "name", "root");
    serializer_setbykeys_format(serial1, "part002", "size", "%d", 380000);
    serializer_dump(serial1, bigbuffer1, sizeof(bigbuffer1));
    printf("DUMP:\n%s\n", bigbuffer1);
    serializer_read(serial2, bigbuffer1);
    serializer_dump(serial2, bigbuffer2, sizeof(bigbuffer2));
    printf("READ:%s\n", bigbuffer2);
    serializer_getbykeys_integer(serial1, "part001", "size", &number1);
    serializer_getbykeys_data(serial1, "part001", "name", string1, sizeof(string1));
    serializer_getbykeys_integer(serial1, "part002", "size", &number2);
    serializer_getbykeys_data(serial1, "part002", "name", string2, sizeof(string2));
    printf("PART1: name=[%s], size=[%ld]\n", string1, (long)number1);
    printf("PART2: name=[%s], size=[%ld]\n", string2, (long)number2);
    serializer_destroy(serial1);
    serializer_destroy(serial2);
    return 0;*/

    // check that datatypes are ok
    assert(sizeof(ciohead)==32);

    // init the lzo library
#ifdef OPTION_LZO_SUPPORT
    if (lzo_init() != LZO_E_OK)
    {
        errprintf("internal error - lzo_init() failed\n");
        exit(EXIT_FAILURE);
    }
#endif // OPTION_LZO_SUPPORT

    // init libgcrypt
    if (crypto_init() != 0)
    {
        errprintf("cannot initialize the crypto environment\n");
        exit(EXIT_FAILURE);
    }

    // init
    options_init();

    if ((g_queue = queue_alloc(FSA_MAX_QUEUESIZE)) == NULL)
    {
        errprintf("cannot initialize g_iobuffer\n");
        exit(EXIT_FAILURE);
    }

    if ((g_iobuffer = iobuffer_alloc(FSA_FEC_IOBUFSIZE, FSA_FEC_VALUE_K * FSA_FEC_PACKET_SIZE)) == NULL)
    {
        errprintf("cannot initialize g_iobuffer (size of buffer: %d bytes)\n", (int)FSA_FEC_IOBUFSIZE);
        exit(EXIT_FAILURE);
    }

    // bulk of the program
    ret = process_cmdline(argc, argv);

    // show status of the operation
    switch (get_status())
    {
        case STATUS_RUNNING:
        case STATUS_FINISHED:
            msgprintf(MSG_FORCE, "\noperation completed\n");
            break;
        case STATUS_ABORTED:
            msgprintf(MSG_FORCE, "\noperation aborted by user\n");
            break;
        case STATUS_FAILED:
            msgprintf(MSG_FORCE, "\noperation failed\n");
            break;
    }

    // cleanup
    iobuffer_destroy(g_iobuffer);
    queue_destroy(g_queue);
    options_destroy();
    crypto_cleanup();

    return !!ret;
}
